python/ray/data/_internal/logical/optimizers.py

```diff
@@ -19,30 +30,211 @@ class PhysicalOptimizer(Optimizer):
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        # Fuse operator names.
+        name = up_op.name + "->" + down_op.name
+
+        down_logical_op = self._op_map.pop(down_op)
+        up_logical_op = self._op_map.pop(up_op)
+
+        # Merge target block sizes.
+        down_target_block_size = down_logical_op._target_block_size
+        up_target_block_size = (
+            up_logical_op._target_block_size
+            if isinstance(up_logical_op, AbstractMap)
+            else None
+        )
+        if down_target_block_size is not None and up_target_block_size is not None:
+            target_block_size = max(down_target_block_size, up_target_block_size)
+        elif up_target_block_size is not None:
+            target_block_size = up_target_block_size
+        else:
+            target_block_size = down_target_block_size
+
+        # Fuse transformation functions.
+        down_transform_fn = down_op.get_transformation_fn()
+        up_transform_fn = up_op.get_transformation_fn()
+
+        def transform_fn(blocks: Iterator[Block]) -> Iterator[Block]:
+            blocks = up_transform_fn(blocks)
+            # TODO(Clark): Add zero-copy batching between transform functions.
+            return down_transform_fn(blocks)
+
+        # We take the downstream op's compute in case we're fusing upstream tasks with a
+        # downstream actor pool (e.g. read->map).
+        compute = get_compute(down_logical_op._compute)
+        ray_remote_args = down_logical_op._ray_remote_args
+        # Make the upstream operator's inputs the new, fused operator's inputs.
+        input_deps = up_op.input_dependencies
+        assert len(input_deps) == 1
+        input_op = input_deps[0]
+
+        # Fused physical map operator.
+        op = MapOperator.create(
+            transform_fn,
+            input_op,
+            name=name,
+            compute_strategy=compute,
+            min_rows_per_bundle=target_block_size,
+            ray_remote_args=ray_remote_args,
+        )
+
+        # Build a map logical operator to be used as a reference for further fusion.
+        # TODO(Clark): This is hacky, remove this once we push fusion to be purely based
+        # on a lower-level operator spec.
+        if isinstance(up_logical_op, AbstractMap):
+            input_op = up_logical_op.input_dependencies[0]
+            fn = up_logical_op._fn
+            fn_args = up_logical_op._fn_args
+            fn_kwargs = up_logical_op._fn_kwargs
+            fn_constructor_args = up_logical_op._fn_constructor_args
+            fn_constructor_kwargs = up_logical_op._fn_constructor_kwargs
+        else:
+            input_op = up_logical_op
+            fn = down_logical_op._fn
+            fn_args = down_logical_op._fn_args
+            fn_kwargs = down_logical_op._fn_kwargs
+            fn_constructor_args = down_logical_op._fn_constructor_args
+            fn_constructor_kwargs = down_logical_op._fn_constructor_kwargs
```

Comment:
Just for my understanding, when this case can happen? Is it for `Read`? And why we set all `fn`-related args from `down_logical_op`, instead of from `up_logical_op`?

Reply:
@c21 Yep it's for when the upstream logical op is a `Read` in which case it won't have any of those `fn`-related args, but I also just realized that this isn't quite right: we should be adopting the downstream op UDF and associated args, not the upstream op, and we should only be taking the upstream op input dependency as the input op (where we bottom-out at the source/read op, which isn't quite correct but should be fine for providing info to physical optimization rules).